pkg/controller/certificaterequests/sync.go

```diff
@@ -0,0 +1,239 @@
+/*
+Copyright 2019 The Jetstack cert-manager contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package certificaterequests
+
+import (
+	"context"
+	"crypto/x509"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/kr/pretty"
+	corev1 "k8s.io/api/core/v1"
+	k8sErrors "k8s.io/apimachinery/pkg/api/errors"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+
+	apiutil "github.com/jetstack/cert-manager/pkg/api/util"
+	"github.com/jetstack/cert-manager/pkg/apis/certmanager/v1alpha1"
+	"github.com/jetstack/cert-manager/pkg/apis/certmanager/validation"
+	"github.com/jetstack/cert-manager/pkg/issuer"
+	logf "github.com/jetstack/cert-manager/pkg/logs"
+	"github.com/jetstack/cert-manager/pkg/util/pki"
+)
+
+const (
+	errorIssuerNotFound    = "IssuerNotFound"
+	errorIssuerNotReady    = "IssuerNotReady"
+	errorIssuerInit        = "IssuerInitError"
+	errorSavingCertificate = "SaveCertError"
+
+	errorCertificateSigningRequestNotFound = "CSRNotFound"
+	errorCertificateSigningRequestParse    = "CSRParseError"
+
+	errorCertificateNotExists = "CertNotExists"
+	errorCertificateParse     = "CertParseError"
+
+	reasonIssuingCertificate = "IssueCert"
+	successCertificateIssued = "CertIssued"
+
+	messageErrorSavingCertificate = "Error saving TLS certificate: "
+)
+
+var (
+	certificateRequestGvk = v1alpha1.SchemeGroupVersion.WithKind("CertificateRequest")
+)
+
+func (c *controller) Sync(ctx context.Context, cr *v1alpha1.CertificateRequest) (err error) {
+	c.metrics.IncrementSyncCallCount(ControllerName)
+
+	log := logf.FromContext(ctx)
+	dbg := log.V(logf.DebugLevel)
+
+	crCopy := cr.DeepCopy()
+	defer func() {
+		if _, saveErr := c.updateCertificateRequestStatus(ctx, cr, crCopy); saveErr != nil {
+			err = utilerrors.NewAggregate([]error{saveErr, err})
+		}
+	}()
+
+	dbg.Info("Fetching existing certificate signing request and certificate from certificate request",
+		"name", crCopy.ObjectMeta.Name)
+	if len(cr.Spec.CSRPEM) == 0 {
+		return errors.New(errorCertificateSigningRequestNotFound)
```

Comment:
Why return an error here? This will cause a retry of the sync, which will result in the same thing until the resource itself is updated. This case should also *never* bit hit as well, as our validation package should ensure that the CSR field is already set.